README.md
A Dart SDK for interacting with a Minecraft server using the RCON protocol.
[Package on pub.dev](https://pub.dev/packages/mc_rcon_dart)

## Features

Provides an API to connect to, log in to, send commands to, and receive data from a Minecraft server via the RCON protocol.

## Getting started

Run `flutter pub add mc_rcon_dart` or `dart pub add mc_rcon_dart`.

Alternatively, add the following to your pubspec.yaml:
```
dependencies:
  mc_rcon_dart: ^1.1.0
```

## Usage
Full code example in [example.dart](example/example.dart).

To import the package:
```dart
import 'package:mc_rcon_dart/mc_rcon_dart.dart';
```

To create the socket (typically want to await on the completion):
```dart
createSocket("172.30.80.31", port: 25575);
```

To set a listener on the socket (note that the handler function must take a List<int> and a String as its only two parameters):
```dart
listen(onData);
```

To log in to the RCON server:
```dart
login("123");
```

To send any command:
```dart
sendCommand("time set 0");
```

To close the connection:
```dart
close();
```

## Additional information

* The RCON documentation is [here](https://wiki.vg/RCON).
* The documentation for Minecraft console commands is [here](https://minecraft.fandom.com/wiki/Commands).
* The documentation for Socket (as of Dart 2.18.1), which is used to communicate with the RCON server, is [here](https://api.dart.dev/stable/2.18.1/dart-io/Socket-class.html).
* Report bugs by making a new issue or send a merge request with the fix, but I'm pretty sure this is all working as is, and I don't expect the RCON protocol to change.

CHANGELOG.md
## 0.0.1
* Implementation of the RCON protocol. Contains full API.

## 0.0.2
* Fixed a typo in the README.

## 1.0.0
* Fixed some debug print lines. Tested package and full API is operational.

## 1.0.1
* Fixed example Dart file name to match pub.dev requirement. (This was a lie.)
* Removed Flutter dependency.

## 1.0.2
* Actually renamed the example file...

## 1.0.3
* Something funky happened with the upload of the last version. Trying again...

## 1.1.0
* Refactored basically every function and split them into many more functions.
* Added a whole lot of comments to go with that refactoring.
* Also added DartDocs to the new functions (but those are all private... so you may never see them :P)
* Modified the example.dart to reflect the new API.
* Modified the README getting started section.
* Modified the README usage section.
* Added Flutter dependency (required for function annotations).
* Added unit tests. THESE ARE NOT DONE. I AM NOT GOING TO WRITE MORE. AT LEAST NOT RIGHT NOW. I DO NOT HAVE THE WILLPOWER TO DO THIS ANY LONGER. Please send help.

test/mc_rcon_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:mc_rcon/mc_rcon.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


test/mc_rcon_dart_test.dart
import 'dart:ffi';
import 'dart:math';

import 'package:flutter/foundation.dart';
import 'package:test/test.dart';
import 'dart:typed_data';
import 'package:mc_rcon_dart/mc_rcon_dart.dart' as mc_rcon;

Random rng = Random();

int lenOfU8L = 10;
int maxU8LLen = 50;
Uint8List errorList = Uint8List(4)
  ..[0] = -1
  ..[1] = -1
  ..[2] = -1
  ..[3] = -1;

void main() {
  testTesterHelpers();
  testU8LCopy4();
  testProcessToInt32();
  testProcessServerResponse();
}

void testTesterHelpers() {
  group("Helper function", () {
    test("_randomU8LLen returns an integer", () {
      // Ensures that _randomU8LLen returns an integer.
      expect(_randomU8LLen(), isA<int>());
    });
    test("_randomU8LLen returns >= 4", () {
      // Ensures that _randomU8LLen returns an integer.
      expect(_randomU8LLen(), greaterThanOrEqualTo(4));
    });
    test("_makeRandomU8L returns lists that are the random len specified", () {
      // Run _makeRandomU8L with random length from _randomU8LLen.
      int randLength = _randomU8LLen();
      List<List<int>> randLists = _makeRandomU8L(randLength);

      // Split the lists to make it cleaner for .length.
      List<int> intList = randLists[0];
      Uint8List u8L = randLists[1] as Uint8List;

      // Ensure that the lengths of the lists == the random length.
      expect(intList.length, equals(randLength));
      expect(u8L.length, equals(randLength));
      expect(intList.length, equals(u8L.length));
    });
    test("_makeRandomU8L returns a List<int> & Uint8List with same ints", () {
      // Make a random len set of lists.
      List<List<int>> randLists = _makeRandomU8L(_randomU8LLen());

      // Split the lists for easier access.
      List<int> intList = randLists[0];
      Uint8List u8L = randLists[1] as Uint8List;

      // Ensure that the lists have the same integers.
      expect(List<int>.from(u8L), equals(intList));
    });
  });
}

void testU8LCopy4() {
  group("u8LCopy4", () {
    test('returns an object of type Uint8List', () {
      // Makes a set of rand len rand int lists and splits for ease of access.
      int randLen = _randomU8LLen();
      List<List<int>> randLists = _makeRandomU8L(randLen);
      Uint8List randUi8L = randLists[1] as Uint8List;

      // Gets the function response.
      Uint8List fnResponse = mc_rcon.u8LCopy4(randUi8L);

      // Ensures that the function returns the 4 integers we know we chose.
      expect(fnResponse, isA<Uint8List>());
    });
    test('returns error list due to a list that is too short.', () {
      // Makes a random len (0-3) list with random integers.
      Uint8List shortUi8L = _makeRandomU8L(rng.nextInt(3))[1] as Uint8List;

      // Gets the function response.
      Uint8List fnResponse = mc_rcon.u8LCopy4(shortUi8L);

      // Ensures that the function returns the error list.
      expect(fnResponse, equals(errorList));
    });
    test('returns error list due to start being too large.', () {
      // Makes a new random len Uint8List.
      int u8LLen = _randomU8LLen();
      Uint8List randU8L = _makeRandomU8L(u8LLen)[1] as Uint8List;

      // Makes a random invalid starting point.
      int randStart = rng.nextInt(3) + (u8LLen - 3);

      // Gets the function response.
      Uint8List fnResponse = mc_rcon.u8LCopy4(randU8L, start: randStart);

      // Ensures that the function response is the error list.
      expect(fnResponse, equals(errorList));
    });
    test('returns error list because neither req is followed.', () {
      // Creates a random Uint8List that is too short.
      int u8LLen = rng.nextInt(3);
      Uint8List randU8L = _makeRandomU8L(u8LLen)[1] as Uint8List;

      // Get a random invalid starting point.
      int randStart = rng.nextInt(3) + (u8LLen - 3);

      // Get the function response.
      Uint8List fnResponse = mc_rcon.u8LCopy4(randU8L, start: randStart);

      // Ensure that the function respondd with the error list.
      expect(fnResponse, equals(errorList));
    });
    test('returns [0,0,0,0] because the source list is empty.', () {
      // Makes a rand len empty Uint8List.
      int randLen = _randomU8LLen();
      Uint8List emptyU8L = Uint8List(randLen);

      // Selects 4 random ints and creates the confirmation list.
      int randSelect = rng.nextInt(randLen - 4);
      Uint8List confirmList = Uint8List(4);
      for (int i = 0; i < 4; i++) {
        confirmList..[i] = 0;
      }

      // Gets the function response.
      Uint8List fnResponse = mc_rcon.u8LCopy4(emptyU8L, start: randSelect);

      // Confirms that the function gave us what we expected.
      expect(fnResponse, equals(confirmList));
    });
    test('returns the 4 copied numbers (full list of 1s).', () {
      // Makes a rand len full of 1s Uint8List.
      int randLen = _randomU8LLen();
      Uint8List randU8L = Uint8List(randLen);
      for (int i = 0; i < randU8L.length; i++) {
        randU8L..[i] = 1;
      }

      // Selects 4 random ints and creates the confirmation list.
      int randSelect = rng.nextInt(randLen - 4);
      Uint8List confirmList = Uint8List(4);
      for (int i = 0; i < 4; i++) {
        confirmList..[i] = randU8L[randSelect + i];
      }

      // Gets the function response.
      Uint8List fnResponse = mc_rcon.u8LCopy4(randU8L, start: randSelect);

      // Confirms that the function gave us what we expected.
      expect(fnResponse, equals(confirmList));
    });
    test('returns the 4 copied numbers (checkerboard start 0).', () {
      // Makes a rand len checkerboard start 0 Uint8List.
      int randLen = _randomU8LLen();
      Uint8List randU8L = Uint8List(randLen);
      for (int i = 0; i < randU8L.length; i++) {
        randU8L..[i] = i % 2;
      }

      // Selects 4 random ints and creates the confirmation list.
      int randSelect = rng.nextInt(randLen - 4);
      Uint8List confirmList = Uint8List(4);
      for (int i = 0; i < 4; i++) {
        confirmList..[i] = randU8L[randSelect + i];
      }

      // Gets the function response.
      Uint8List fnResponse = mc_rcon.u8LCopy4(randU8L, start: randSelect);

      // Confirms that the function gave us what we expected.
      expect(fnResponse, equals(confirmList));
    });
    test('returns the 4 copied numbers (checkerboard start 1).', () {
      // Makes a rand len checkerboard start 0 Uint8List.
      int randLen = _randomU8LLen();
      Uint8List randU8L = Uint8List(randLen);
      for (int i = 0; i < randU8L.length; i++) {
        if (i % 2 == 0) {
          randU8L..[i] = 1;
        } else {
          randU8L..[i] = 0;
        }
      }

      // Selects 4 random ints and creates the confirmation list.
      int randSelect = rng.nextInt(randLen - 4);
      Uint8List confirmList = Uint8List(4);
      for (int i = 0; i < 4; i++) {
        confirmList..[i] = randU8L[randSelect + i];
      }

      // Gets the function response.
      Uint8List fnResponse = mc_rcon.u8LCopy4(randU8L, start: randSelect);

      // Confirms that the function gave us what we expected.
      expect(fnResponse, equals(confirmList));
    });
    test('returns the 4 copied numbers (full random list).', () {
      // Makes a set of rand len rand int lists and splits for ease of access.
      int randLen = _randomU8LLen();
      List<List<int>> randLists = _makeRandomU8L(randLen);
      List<int> randomInts = randLists[0];
      Uint8List randUi8L = randLists[1] as Uint8List;

      // Gets a random set of 4 integers from the random ints.
      int randSelect = rng.nextInt(randLen - 4);
      Uint8List confirmList = Uint8List(4);
      for (int i = 0; i < 4; i++) {
        confirmList..[i] = randomInts[randSelect + i];
      }

      // Gets the function response.
      Uint8List fnResponse = mc_rcon.u8LCopy4(randUi8L, start: randSelect);

      // Ensures that the function returns the 4 integers we know we chose.
      expect(fnResponse, equals(confirmList));
    });
  });
}

void testProcessToInt32() {
  group("processToInt32", () {
    test("returns an object of type int", () {
      // Make a random len random integer Uint8List
      int randLen = _randomU8LLen();
      Uint8List randU8L = _makeRandomU8L(randLen)[1] as Uint8List;

      // Select a random valid starting point.
      int randStart = rng.nextInt(randLen - 4);

      // Get the function response.
      int fnResponse = mc_rcon.processToInt32(randU8L, randStart);

      // Ensure that function gave us an int.
      expect(fnResponse, isA<int>());
    });
    test("returns -1 because less than 4 Uint8s were provided.", () {
      // Make a random len random integer Uint8List
      int randLen = rng.nextInt(3);
      Uint8List randU8L = _makeRandomU8L(randLen)[1] as Uint8List;

      // Get the function response.
      int fnResponse = mc_rcon.processToInt32(randU8L, 0);

      // Ensure that function gave us an int.
      expect(fnResponse, isA<int>());
    });
  });
}

void testProcessServerResponse() {
  group("processServerResponse", () {
    test("calls the handler function with a List<int> and String", () {
      dynamic fnHeaders = [];
      dynamic fnPayload = "";

      Uint8List msg = mc_rcon.createMessage(10, 1, "");
      mc_rcon.processServerResponse(msg, (List<int> headers, String payload) {
        fnHeaders = headers;
        fnPayload = payload;
      });

      expect(fnHeaders, isA<List<int>>());
      expect(fnPayload, isA<String>());

      Uint8List msgNum2 = mc_rcon.cM(10, 1, "");
      mc_rcon.pSR(msgNum2, (List<int> headers, String payload) {
        fnHeaders = headers;
        fnPayload = payload;
      });

      expect(fnHeaders, isA<List<int>>());
      expect(fnPayload, isA<String>());
    });
    dynamic fnHeaders = [];
    test("fails because bad/no auth", () {
      dynamic fnPayload = "";

      Uint8List msg = mc_rcon.createMessage(10, 99999, "", -1);
      mc_rcon.processServerResponse(msg, (List<int> headers, String payload) {
        fnHeaders = headers;
        fnPayload = payload;
      });

      expect(fnHeaders, isEmpty);
      expect(fnPayload, isEmpty);
    });
    test("fails because unknown request ID", () {
      dynamic fnHeaders = [];
      dynamic fnPayload = "";

      Uint8List msg = mc_rcon.createMessage(
          10, 99999, "", rng.nextInt(99999999) - rng.nextInt(99999));
      mc_rcon.processServerResponse(msg, (List<int> headers, String payload) {
        fnHeaders = headers;
        fnPayload = payload;
      });

      expect(fnHeaders, isEmpty);
      expect(fnPayload, isEmpty);
    });
  });
}

int _randomU8LLen() {
  return rng.nextInt(maxU8LLen) + 5;
}

List<List<int>> _makeRandomU8L(int length) {
  List<int> randomInts = [];
  while (randomInts.length < length) {
    randomInts.add(rng.nextInt(255));
  }

  Uint8List uint8list = Uint8List(length);
  for (int i = 0; i < uint8list.length; i++) {
    uint8list..[i] = randomInts[i];
  }

  return [randomInts, uint8list];
}


lib/src/rcon_api.dart
import 'dart:io';
import 'dart:typed_data';

import 'rcon_vars.dart';
import 'rcon_helpers.dart';

/// Creates and stores a socket connected to the RCON server
/// with the given host (IP/FQDN) and port. Port defaults to
/// 25575 if no port is specified.
Future<void> createSocket(String host, {int port = 25575}) async {
  // Creates the socket by connecting the socket to the specified
  // host and port.
  rconSck = await Socket.connect(host, port);
}

/// Closes the socket to the RCON server. Returns a bool that
/// specified whether the socket was successfully destroyed.
bool close() {
  // Checks to ensure that the RCON socket exists.
  if (rconSck == null) {
    return false;
  }

  // Destroys the socket, which also closes the connection.
  rconSck!.destroy();

  return true;
}

/// Send a message with the given message ID and message payload.
/// Returns a boolean that specifies if the command was successful.
bool sendMsg(int msgID, String payload) {
  // Ensures that the RCON socket exists.
  if (rconSck == null) {
    return false;
  }

  // Message length is the payload length + 10 to account
  // for the headers and suffix.
  int msgLen = 10 + payload.length;

  // Creates the full RCON message.
  Uint8List fullMsg = cM(msgLen, msgID, payload);

  // Add the RCON message to the socket stream.
  rconSck!.add(fullMsg);
  print("mc_rcon: sent payload ($fullMsg) on socket");

  return true;
}

/// Log in to the RCON server using the given socket and password.
/// Returns a boolean that specifies if the command was successful.
bool login(String password) {
  // Sends an RCON message with request ID = 3 (authenticate)
  // with the password as the payload.
  return sendMsg(3, password);
}

/// Send the provided command to the RCON server using the
/// Returns a boolean that specifies if the command was successful.
bool sendCommand(String command) {
  // Sends an RCON message with request ID = 2 (command)
  // with the String command as the payload.
  return sendMsg(2, command);
}

/// Starts listening on the socket for packets sent by the RCON
/// server. Returns a boolean that specifies if the socket has
/// started listening. Note: onData must accept a List<int> and
/// a String as the only parameters.
bool listen(Function onData) {
  // Checks to ensure that the RCON socket exists.
  if (rconSck == null) {
    return false;
  }

  // Starts listening on the RCON socket.
  // Calls the first handler if we receive data, calls onError
  // if there is an error on the stream, calls onDone when the
  // client or the server ends the connection.
  rconSck!.listen(
    (Uint8List data) {
      pSR(data, onData);
    },
    onError: (error) {
      print('mc_rcon: Error with the connection to the server: $error');
      rconSck!.destroy();
    },
    onDone: () {
      print('mc_rcon: The server has ended the connection.');
      rconSck!.destroy();
    },
    cancelOnError: false,
  );

  return true;
}


lib/src/rcon_helpers.dart
import 'dart:typed_data';
import 'package:flutter/foundation.dart';

import 'rcon_vars.dart';

/// Returns a Uint8Lit that contains only 4 integers starting
/// from original[start]. Start defaults to 0 if not set. If
/// 4 integers cannot be copied from the list (e.g. list length
/// is less than 4 or start is too large to copy 4 integers), a
/// list of [-1, -1, -1, -1] will be returned. Returns 0 in place
/// of any integer if that location is empty in the source list.
Uint8List _u8LCopy4(Uint8List original, {int start = 0}) {
  // Ensure that the Uint8List is at least len 4 and that
  // start is list.length-4 or less
  if (original.length < 4 || start + 3 > original.length - 1) {
    // If not, return a list that is all -1.
    Uint8List errorList = Uint8List(4)
      ..[0] = -1
      ..[1] = -1
      ..[2] = -1
      ..[3] = -1;

    return errorList;
  }

  // Creates a new length 4 Uint8List and sets each of the
  // values to a value from the original list.
  Uint8List copiedUint8s = Uint8List(4)
    ..[0] = original[start]
    ..[1] = original[start + 1]
    ..[2] = original[start + 2]
    ..[3] = original[start + 3];

  return copiedUint8s;
}

/// Returns a Uint8Lit that contains only 4 integers starting
/// from original[start]. Start defaults to 0 if not set. If
/// 4 integers cannot be copied from the list (e.g. list length
/// is less than 4 or start is too large to copy 4 integers), a
/// list of [-1, -1, -1, -1] will be returned. Returns 0 in place
/// of any integer if that location is empty in the source list.
@visibleForTesting
Uint8List u8LCopy4(Uint8List original, {int start = 0}) {
  return _u8LCopy4(original, start: start);
}

/// Returns a little Endian int32 interpreted from 4 Uint8s
/// retrieved from data (Uint8List) starting at an index of
/// start (int). The Uint8list length must be at least 4 and
/// start must be data.length - 4 or smaller. Returns -1
/// if processing has failed.
int _processToInt32(Uint8List data, int start) {
  // Checks to ensure that the Uint8List is longer than 4
  // and that the starting index is valid to parse out a
  // 32-bit integer.
  if (data.length < 4 || start > (data.length - 4)) {
    return -1;
  }

  // Copies the 4 uint8s we need to parse the int32.
  Uint8List copiedUint8s = _u8LCopy4(data, start: start);

  // I don't know what this does. But it gives me access
  // to the list and the ability to parse out a 32-bit int.
  var blob = ByteData.sublistView(copiedUint8s);

  // Process out the 32-bit integer assuming little Endian
  // (which is what the RCON protocol uses).
  int processedInt32 = blob.getInt32(0, Endian.little);

  return processedInt32;
}

/// Returns a little Endian int32 interpreted from 4 Uint8s
/// retrieved from data (Uint8List) starting at an index of
/// start (int). The Uint8list length must be at least 4 and
/// start must be data.length - 4 or smaller. Returns -1
/// if processing has failed.
@visibleForTesting
int processToInt32(Uint8List data, int start) {
  return _processToInt32(data, start);
}

/// Returns a List<int> that contains the 3 header integers
/// processed from the Uint8List of data.
List<int> _processHeaders(Uint8List data) {
  // Processes out each of the 3 header integers.
  int msgLength = _processToInt32(data, 0);
  int respReqID = _processToInt32(data, 4);
  int commandID = _processToInt32(data, 8);

  // Creates a list made of the header integers.
  List<int> msgHeader = [msgLength, respReqID, commandID];

  return msgHeader;
}

/// Returns a List<int> that contains the 3 header integers
/// processed from the Uint8List of data.
// @visibleForTesting
// List<int> processHeaders(Uint8List data) {
//   return _processHeaders(data);
// }

/// Returns a boolean that represents whether the response ID
/// represents a good packet. Good packet means response
/// ID == original request ID. Bad auth packet means response
/// ID == -1.
bool _processResponseID(int respID) {
  if (respID == requestID) {
    // If the response ID == original request ID,
    // we received a good packet.
    print("mc_rcon: Good packet received.");
    return true;
  } else if (respID == -1) {
    // If the response ID == -1, we haven't authenticated
    // properly, which can mean we sent the wrong password
    // or we haven't authenticated yet.
    print(
        "mc_rcon: Bad authentication. Incorrect password or you haven't logged in yet.");
    return false;
  } else {
    // Catch-all for all other response IDs. Should never trigger.
    print("mc_rcon: Received unknown request ID.");
    return false;
  }
}

/// Returns a boolean that represents whether the response ID
/// represents a good packet. Good packet means response
/// ID == original request ID. Bad auth packet means response
/// ID == -1.
// @visibleForTesting
// bool processResponseID(int respID) {
//   return _processResponseID(respID);
// }

/// Processes the server response (represented as a Uint8List)
/// and calls onData handler if we have received a good packet.
void _processServerResponse(Uint8List data, Function onData) {
  // Parses out the message headers and payload.
  List<int> rconHeaders = _processHeaders(data);
  String payload = String.fromCharCodes(data, 12);

  // Pulls out the messsage length to ensure the integrity
  // of the message and the response ID to print.
  int messageLen = rconHeaders[0];
  int responseID = rconHeaders[1];
  print("mc_rcon: Server response id: $responseID");

  // Ensures that the data we recieved is the same
  // as what the length of the message is supposed to be.
  bool badMessage = (data.length == messageLen);

  // Sends the headers and payload to the user handler function
  // if the response is good (we receive our own request ID).
  if (!badMessage && _processResponseID(responseID)) {
    onData(rconHeaders, payload);
  }
}

/// Processes the server response (represented as a Uint8List)
/// and calls onData handler if we have received a good packet.
@visibleForTesting
void processServerResponse(Uint8List data, Function onData) {
  return _processServerResponse(data, onData);
}

/// Processes the server response (represented as a Uint8List)
/// and calls onData handler if we have received a good packet.
@protected
void pSR(Uint8List data, Function onData) {
  return _processServerResponse(data, onData);
}

/// Sets every int in dataList to the Uint32List using a ByteData buffer.
void _setUint32s(Uint32List int32List, List<int> dataList) {
  // Views the buffer of the Uint32list.
  ByteData bd = ByteData.view(int32List.buffer);

  // Used to offset the bytes assigned to the ByteData.
  // Otherwise we will overwrite the already written bytes.
  int bdByteOffset = 0;

  // Loops through every int data and assigns it as a little
  // Endian Uint32 to the Uint32list's ByteData buffer.
  for (int data in dataList) {
    bd.setUint32(bdByteOffset, data, Endian.little);
    bdByteOffset += 4;
  }
}

/// Sets every int in dataList to the Uint32List using a ByteData buffer.
// @visibleForTesting
// void setUint32s(Uint32List int32List, List<int> dataList) {
//   return _setUint32s(int32List, dataList);
// }

/// Returns a Uint8List that represents the header of
/// the RCON message. Assembles the header with the
/// specified length and message ID (i.e. type of message).
Uint8List _assembleHeader(int payloadLen, int msgID, [int? overrideReqID]) {
  // Creates a new, length 3, Uint32List.
  Uint32List headerAs32 = Uint32List(3);

  // Sets the three Uint32s to the proper header integers.
  _setUint32s(headerAs32, [
    payloadLen,
    overrideReqID != null ? overrideReqID : requestID,
    msgID,
  ]);

  // Transforms the header Uint32List into a Uint8List
  // for transmission.
  Uint8List header = headerAs32.buffer.asUint8List();

  return header;
}

/// Returns a Uint8List that represents the header of
/// the RCON message. Assembles the header with the
/// specified length and message ID (i.e. type of message).
// @visibleForTesting
// Uint8List assembleHeader(int payloadLen, int msgID) {
//   return _assembleHeader(payloadLen, msgID);
// }

/// Returns a Uint8List that represents the suffix of
/// the RCON message. The suffix is two NULLs in ASCII,
/// one to end the payload, and one to end the message.
Uint8List _assembleSuffix() {
  // Creates a new, length 2, Uint8List.
  Uint8List suffix = Uint8List(2);

  // Views the buffer of the list so we can add to the list.
  ByteData suffixBD = ByteData.view(suffix.buffer);

  // Adds the two NULLs to the Uint8List.
  suffixBD.setUint8(0, 0);
  suffixBD.setUint8(1, 0);

  return suffix;
}

/// Returns a Uint8List that represents the suffix of
/// the RCON message. The suffix is two NULLs in ASCII,
/// one to end the payload, and one to end the message.
// @visibleForTesting
// Uint8List assembleSuffix() {
//   return _assembleSuffix();
// }

/// Assembles a list of Uint8Lists into one Uint8List.
Uint8List _assembleUint8Lists(List<Uint8List> msgParts) {
  // Creates a new BytesBuilder to assemble the ints.
  BytesBuilder bBuilder = BytesBuilder();

  // Adds every list of Uint8s to the BB.
  // This automatically compiles the list of ints.
  for (Uint8List part in msgParts) {
    bBuilder.add(part);
  }

  // Returns the BB's data as a Uint8list.
  return bBuilder.toBytes();
}

/// Assembles a list of Uint8Lists into one Uint8List.
// @visibleForTesting
// Uint8List assembleUint8Lists(List<Uint8List> msgParts) {
//   return _assembleUint8Lists(msgParts);
// }

/// Returns the whole RCON message as a Uint8List. Requires the
/// message length, message ID, and the payload.
Uint8List _createMessage(int msgLen, int msgID, String payload,
    [int? overrideReqID]) {
  // Example login data: (u__ refers to the type of unsigned integer)
  // 0d00 0000 | dec0 ad0b | 0300 0000 | 3132 3300 00
  //  len u32  |  pID u32  | cmdID u32 | payload/password u8

  // Creates the 3 parts of the message using their respective
  // constructors (2 are self-written, one built into Uint8List).
  Uint8List header = _assembleHeader(msgLen, msgID, overrideReqID);
  Uint8List msgAsIntList = Uint8List.fromList(payload.codeUnits);
  Uint8List suffix = _assembleSuffix();

  // Assembles the 3 parts into 1 Uint8List to return;
  Uint8List fullMsg = _assembleUint8Lists([header, msgAsIntList, suffix]);

  return fullMsg;
}

/// Returns the whole RCON message as a Uint8List. Requires the
/// message length, message ID, and the payload.
@visibleForTesting
Uint8List createMessage(int msgLen, int msgID, String payload,
    [int? overrideReqID]) {
  return _createMessage(msgLen, msgID, payload, overrideReqID);
}

/// Returns the whole RCON message as a Uint8List. Requires the
/// message length, message ID, and the payload.
@protected
Uint8List cM(int msgLen, int msgID, String payload) {
  return _createMessage(msgLen, msgID, payload);
}


lib/mc_rcon.dart
library mc_rcon;

import 'dart:io';
import 'dart:math';
import 'dart:typed_data';

import 'package:flutter/foundation.dart';

/// The socket that is used to communicate with the RCON server.
/// Generated when createSocket() is run.
Socket? rconSck;

/// The randomly generated request ID that is sent with every message
/// to the RCON server. Used to ensure that the commands sent to and
/// received from the server are ours, and not another user's.
int requestID = Random().nextInt(2147483647);

/// Creates and stores a socket connected to the RCON server
/// with the given host (IP/FQDN) and port.
Future<void> createSocket(String host, {int port = 25575}) async {
  rconSck = await Socket.connect(host, port);
}

/// Closes the socket to the RCON server. Returns a bool that specified
/// whether the socket was successfully destroyed.
bool close() {
  if (rconSck == null) {
    return false;
  }
  rconSck!.destroy();
  return true;
}

/// Starts listening on the socket for packets sent by the RCON server.
/// Returns a boolean that specifies if the socket has started listening.
/// Note: onData must accept a Uint8List as the only parameter. The payload
/// of the message starts at element 12.
bool listen(Function onData) {
  if (rconSck == null) {
    return false;
  }

  rconSck!.listen(
    (Uint8List data) {
      Uint8List respReqIDInts = Uint8List(4)
        ..[0] = data[4]
        ..[1] = data[5]
        ..[2] = data[6]
        ..[3] = data[7];
      var blob = ByteData.sublistView(respReqIDInts);
      int respReqID = blob.getInt32(0, Endian.little);
      if (kDebugMode) {
        print("response id: $respReqID");
      }

      if (respReqID == requestID) {
        if (kDebugMode) {
          print("mc_rcon: Good packet received.");
        }
        onData(data);
      } else if (respReqID == -1) {
        if (kDebugMode) {
          print(
              "mc_rcon: Bad authentication. Incorrect password or you haven't logged in yet.");
        }
      } else {
        if (kDebugMode) {
          print("mc_rcon: Received unknown request ID.");
        }
      }
    },
    onError: (error) {
      if (kDebugMode) {
        print('mc_rcon: Error with the connection to the server: $error');
      }
      rconSck!.destroy();
    },
    onDone: () {
      if (kDebugMode) {
        print('mc_rcon: The server has ended the connection.');
      }
      rconSck!.destroy();
    },
    cancelOnError: false,
  );

  return true;
}

/// Send a message with the given message ID and message payload.
/// Returns a boolean that specifies if the command was successful.
bool sendMsg(int msgID, String msg) {
  if (rconSck == null) {
    return false;
  }

  // Example login data: (u__ refers to the type of unsigned integer)
  // 0d00 0000 | dec0 ad0b | 0300 0000 | 3132 3300 00
  //  len u32  |  pid u32  | cmdid u32 | payload/password u8

  int payloadLen = 10 + msg.length;

  Uint32List headerAs32 = Uint32List(3);
  var headerBD = ByteData.view(headerAs32.buffer);
  headerBD.setUint32(0, payloadLen, Endian.little);
  headerBD.setUint32(4, requestID, Endian.little);
  headerBD.setUint32(8, msgID, Endian.little);
  Uint8List header = headerAs32.buffer.asUint8List();

  Uint8List passwordAsIntList = Uint8List.fromList(msg.codeUnits);

  Uint8List suffix = Uint8List(2);
  var suffixBD = ByteData.view(suffix.buffer);
  suffixBD.setUint8(0, 0);
  suffixBD.setUint8(1, 0);

  BytesBuilder bBuilder = BytesBuilder();
  bBuilder.add(header);
  bBuilder.add(passwordAsIntList);
  bBuilder.add(suffix);
  Uint8List payload = bBuilder.toBytes();

  rconSck!.add(payload);
  if (kDebugMode) {
    print("mc_rcon: sent payload ($payload) on socket");
  }

  return true;
}

/// Log in to the RCON server using the given socket and password.
/// Returns a boolean that specifies if the command was successful.
bool login(String password) {
  return sendMsg(3, password);
}

/// Send the provided command to the RCON server using the
/// Returns a boolean that specifies if the command was successful.
bool sendCommand(String message) {
  return sendMsg(2, message);
}
